File: one-more-time/Lagrange/Lagrange.hs
import Text.Printf
import System.IO

data Point = Point { x :: Double, y :: Double }

zeros :: Int -> [Double]
zeros n = replicate n 0.0

denom :: Int -> [Point] -> Double
denom i points = foldl (\p j -> if i /= j then p * (x (points !! i) - x (points !! j)) else p) 1.0 [0..length points - 1]

interpolate :: Int -> [Point] -> [Double]
interpolate i points = let
    coeff = zeros (length points)
    coeff' = 1 / denom i points : map (\k -> if k == i then 0.0 else 1 / denom i points) [0..length points - 1]
    newCoeff = foldl (\newCoeff k -> if k /= i then
        let
            x' = x (points !! k)
            newCoeff' = zipWith (\j c -> if j < 2 then c + newCoeff !! j else c) [0..length newCoeff - 1] newCoeff
            newCoeff'' = zipWith (\j c -> if j < 2 then c - x' * newCoeff !! (j+1) else c) [0..length newCoeff - 1] newCoeff
        in newCoeff' else newCoeff) coeff' [0..length points - 1]
    in newCoeff

lagrange :: [Point] -> [Double]
lagrange points = let
    polynom = zeros (length points)
    coeff = zeros (length points)
    coeff' = interpolate 0 points
    polynom' = foldl (\polynom i -> let
        coeff'' = interpolate i points
        polynom'' = zipWith (\k c -> c + y (points !! i) * coeff !! k) [0..length coeff - 1] polynom
        in polynom'') polynom [0..length points - 1]
    in polynom'

main :: IO ()
main = do
    contents <- readFile "ingput.txt"
    let nums = map read $ words contents
    let degree = head nums + 1
    let pointPairs = chunksOf 2 $ tail nums
    let points = map (\[x, y] -> Point x y) pointPairs
    let finalCoeff = lagrange points
    writeFile "otput.txt" $ unwords $ map (formatDouble 2) finalCoeff
    putStrLn "Lagrange coefficients successfully written."

chunksOf :: Int -> [a] -> [[a]]
chunksOf _ [] = []
chunksOf n xs = take n xs : chunksOf n (drop n xs)

formatDouble :: Int -> Double -> String
formatDouble n = printf ("%." ++ show n ++ "f")

File: one-more-time/Lagrange/Lagrange.js
function zeros(n) {
   var arr = new Array(n);
   for (var i=n; i--;) {
     arr[i] = 0;
   }
   return arr;
}

function denom(i, points) {
   var p = 1;
   var x = points[i].x;
   for (var j = points.length; j--;) {
      if (i != j) {
        p *= x - points[j].x;
      }
   }
   return p;
}

// interpolating in form of Li polynomial
function interpolate(i, points) {
   var coeff = zeros(points.length);
   coeff[0] = 1/denom(i,points);
   var new_coeff;

   for (var k = 0; k < points.length; k++) {
        if (k == i) {
          continue;
        }
        new_coeff = zeros(points.length);
        for (var j= (k < i) ? k+1 : k; j--;) {
            if (j < 2) {
                new_coeff[j+1] += coeff[j];
                new_coeff[j] -= points[k].x*coeff[j];
                //console.log(k);
                //console.log(j);
                //console.log(new_coeff);
            }
        }
        coeff = new_coeff;
   }
   return coeff;
}

// coefficients of polynomial
function Lagrange(points) {
   var polynom = zeros(points.length);
   var coeff;
   for (var i=0; i<points.length; i++) {
     coeff = interpolate(i, points);
     for (var k=0; k<points.length; k++) {
        polynom[points.length - 1 - k] += points[i].y*coeff[k];
     }
   }
   return polynom;
}

function makeStruct(keys) {
  if (!keys) return null;
  const k = keys.split(', ');
  const count = k.length;

  /** @constructor */
  function constructor() {
    for (let i = 0; i < count; i++) this[k[i]] = arguments[i];
  }
  return constructor;
}

const point = new makeStruct("x, y");
var points = [];

var fs = require('fs')
fs.readFile('ingput.txt', 'utf-8', (err, data) => {
   if (err) throw err;
      let info = data.toString().match(/\d+/g);
   
      for (var e = 0; e <= info[0]; e++) {
         points.push(new point(info[2*e + 1], info[2*e + 2]));
      }
      const final = Lagrange(points).toString();
      console.log(final);
      var save = require('fs');
      save.writeFile('otput.txt', final, (err) => {
         if (err) throw err;
         else{
            console.log("Lagrange coefficients successfully written.");
         }
   });
});